### ISSUE TITLE
Proposal to reduce the load on the web server with DataURI

### ISSUE THREAD

**Issue by narodmon** [phpmyadmin/phpmyadmin#15362]
Good day colleagues.
 
Let me offer you a fairly simple way to reduce the load on the web server when using PMA just to replace a lot of small graphics with DataURI (inline base64 form of the mediadata instead of separate files).
 
The problem lies in the fact that the PMA, on its own load, creates a huge number of threads for downloading data from the server and often falls under load restrictions. In the case of DataURI the number of streams, it will be possible to reduce several times by replacing the entire small graphics with its base64 inline representation in the http / php code.
 
Thank you for your understanding, I hope you will listen to the wishes of your client and the founder of a non-commercial Project in the field of the Internet of things.

**Comment by ibennetch** [comment]
Hello and thank you for taking the time to write to us with detail.

Could you share with us a bit more information about the root problem that
has caused you to come up with this idea? Web server configuration,
anything special about the server, etc? Which phpMyAdmin version?

This problem is particularly interesting because we use CSS sprites to
reduce the number of HTTP requests and as far as I know you shouldn’t be
seeing this problem under a normal configuration.

Thanks!

**Comment by narodmon** [comment]
phpMyAdmin 4.9.0.1 on nginx/1.16.0 + php-fpm
Steps for more details:
1. Pls open PMA in (eg) FireFox with opened tab "Network" in developers tools.
2. Press Ctrl+F5 in FireFox (full page reload).
3. I can see 73(!) requests (1 html, 6 css, 27 js, 6 xhr, 33 images..) in first second to my webserver only from 1(!) user session.
PS Also 27 jscripts its also not a good idea too.
PPS More than 8-10 requests/sec per IP its a result of bad website optimization.

**Comment by OlafvdSpek** [comment]
> PPS More than 8-10 requests/sec per session its a result of bad website optimization.

Says who? ;)
10+ requests for a page seems normal nowadays.

**Comment by narodmon** [comment]
If you are the one user on your site - its OK. But if your site the is the production site....
Each PMA user creates load on the web server as 10-20 average users of the site, do you think this is normal ?
Please, respect the work of our colleagues, system administrators and save the budget of their clients without wasting server resources.

**Comment by narodmon** [comment]
I have too much messages "limiting requests, excess .." in my nginx error log file mostly for PMA alias and spammers.

**Comment by OlafvdSpek** [comment]
> Each PMA user creates load on the web server as 10-20 average users of the site, do you think this is normal ?

Yes
Note most of the resources should get cached, so future requests should cause less load.

**Comment by narodmon** [comment]
Yes, the next load is 50% less (if client side caching is enabled), but the initial PMA load causes an error HTTP 429 (Too Many Connections).
Yes, of course, for the sake of PMA alone, you can increase the limits for one session (not one server), but is it worth it to increase the attacks of DDoS attacks? Definitely not.
Well, it would be if this problem was resolved somehow very difficult, but after all, it is for these purposes that both invented css sprites and datauri.
So, if you are too lazy to help a huge number of your users (although most of them certainly don’t know, and where the nginx logs are stored and analyzed), please consider my offer in future versions of PMA. This is very easy done with the php, even the folder structure in the distribution does not need to change, but you need to generate DataURI on the fly or take it from the database.

**Comment by narodmon** [comment]
Thanks for your time, I will wait for new optimized versions of PMA.

**Comment by OlafvdSpek** [comment]
> Yes, the next load is 50% less (if client side caching is enabled), but the initial PMA load causes an error HTTP 429 (Too Many Connections).

What limit did you pick?
Seems to be lower than the limit browsers are expecting.

**Comment by narodmon** [comment]
At this moment PMA works well only if limit_req ~120 per ip per sec, eg for IoT project 30 is enough (per IP per sec too) w/o errors in log file.
For the browser normally up to 8-16 connections, mostly 4.
